### PR TITLE
fix: Fix UN & AT VR processing logic (#167)

### PR DESCRIPTION
**Pull request by WoonchanCho**
* Change UnknownValue (UN) to be extended from BinaryRepresentation
* Add the writeTwoUint16s method to BufferStream to handle to process AT
* Optimize the readString method of BufferStream
* Adds test case to check the equality of pre-write and post-write buffer

### CODE REVIEW

**Comment by WoonchanCho** [comment]
Hi @pieper,
This is a fix for the issue #167.

As you mentioned earlier, several issues together made the bigger issue.
This is the summary of the issue and fix.

1) The situation: 
  I found that the file size keeps growing when iterating reading and writing images by using the previous output as the input of the next iteration.
  I found that only some of the UN types elements (not all) grew in size at each iteration.
  Each iteration took took much time and even worse, the iteration process failed only at the fifth iteration of 100 iterations with the attached dcm file.[dcm.zip](https://github.com/dcmjs-org/dcmjs/files/5938424/dcm.zip)

2) The cause
  a) file growth : UN type is being handled as a string representation while this should be binary.
  b) Slowness & heap out of expection: There were inefficient codes at handling big string text. 
        ` while (index++ < numOfMulti) {
            var charCode = this.readUint8();
            string += String.fromCharCode(charCode);
        }`
        The above code unnecessary creates 1, 2, 3, ... nth string. 
        This will be okay at small text, but in case of big text, it consumes too many heap memory (garbage collection takes too much time) and makes program fail.

3) Fixes
    a) Change UnknownValue (UN) to be extended from BinaryRepresentation
    b) Add the writeTwoUint16s method to BufferStream to handle to process AT
        => This is a separate issue, I found that the AT handling logic is not correct.
    c) Optimize the readString method of BufferStream
    d) Add test case to check the equality of pre-write and post-write buffer
    e) Add Iterate-test.js into test folder

After this fix, the 100 iteration test with the above file finished in 3 secs with much less heap used.

You can test it too if you want.

Thank you.

File: src/ValueRepresentation.js [hunk 1]
```diff
@@ -955,17 +955,17 @@ class UniversalResource extends StringRepresentation {
...
         this.padByte = "00";
         this.noMultiple = true;
     }
 
-    readBytes(stream, length) {
-        return stream.readString(length);
-    }
+    // readBytes(stream, length) {
+    //     return stream.readString(length);
+    // }
```
Comment (pieper) [inline src/ValueRepresentation.js:968]: Better to remove completely rather than leave around dead code.

File: test/iteration-test.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+async function usage(title, i) {
+    console.log(`---------- [${i}] ${title} ----------`)
+    const dict = process.memoryUsage();
+    Object.keys(dict).forEach(key => {
+        const used = dict[key] / 1024 / 1024;
+        console.log(`${key}: ${Math.round(used * 100) / 100} MB`);
+    });
+}
+
+iterate(inputPath);
```
Comment (pieper) [inline test/iteration-test.js:43]: Is this used anywhere?  Can it be made into a test?

**Review by pieper — CHANGES_REQUESTED** [review]
Thanks for working on this 👍   The fix looks good to me.

Would it be possible for you to put your file.dcm.zip as a release asset in the data repository and set up the iterate as part of the test suite? https://github.com/dcmjs-org/data  That will help ensure this doesn't pop up again.

**Comment by pieper** [comment]
Thanks - is it ready to merge?

**Comment by WoonchanCho** [comment]
@pieper Yes, you can merge now.

dcm.zip is the one I found on the internet: http://www.dclunie.com/pixelmedimagearchive/upmcdigitalmammotomocollection/index.html
I just emailed the owner of the images asking if one of those files can be used for this open source project.
I will add dcm.zip once he has approved.

Thank you.

**Comment by WoonchanCho** [comment]
@pieper, this is not related to this issue. I have some questions (not an issue all) on the library.
What will be the best place to ask my questions?

**Comment by pieper** [comment]
Thank you @WoonchanCho 👍   

If the data is from David Clunie he's very likely to approve of using it for this purpose.  I talk with him frequently on another project so if you don't hear back from him just ping me and I can ask.

Regarding any questions about his library, I guess the easiest is to file an issue in the repo here.

**Comment by WoonchanCho** [comment]
@pieper, oh, it's small world, I just randomly googled it to find a breast tomo image :)
I will do so unless I've not heard from hi.

Thank you so much!👍

**Pull request merged**